LanguageFeatures/Extension-types/static_analysis_extension_types_A20_t04.dart

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion The extension type erasure of an extension type V is obtained by
+/// recursively replacing every subterm of V which is an extension type by the
+/// corresponding representation type.
+///
+/// Let X1 extends B1, .. Xs extends Bs be a declaration of the type parameters
+/// of a generic entity (it could be a generic class, extension type, mixin,
+/// typedef, or function). Let BBj be the extension type erasure of
+/// Bj, for j in 1 .. s. It is a compile-time error if
+/// X1 extends BB1, .. Xs extends BBs has any compile-time errors.
+///
+/// @description Check that the erasure of an extension type `V` is obtained by
+/// recursively replacing every subterm of `V` which is an extension type by the
+/// corresponding representation type. Test function types
+/// @author sgrekhov22@gmail.com
+
+// SharedOptions=--enable-experiment=inline-class
+
+extension type const Check._(Object? _) {
+  const Check(Object? object, {required Object? expected}) : _ = object,
+      assert(identical(expected, object));
+}
+
+typedef GenFunction<T> = X Function<X>(X x);
```

Comment:
I don't think it's useful to test the behavior of a type alias that doesn't use its type argument(s), that would be tested already in various type alias tests.

So the tests would be just as good (and it has exactly the same meaning after alias expansion) if the type aliases don't take any type arguments in the first place:

```suggestion
typedef GenFunction = X Function<X>(X x);
```

This means that this test will test a generic function type against an extension type whose representation type is that same generic function type. That's fine (but I think some of the test cases will be identical after the unused type arguments have been eliminated, and we should only have one of each).

However, I think the main thing which was missing is still missing (or do we have it in some other tests?): Function types where extension types occur as subterms:

```dart
extension type ExtInt(int it) {}

typedef F1 = ExtInt Function(ExtInt arg1, [List<ExtInt> arg2]);
typedef F2 = int Function(int arg1, [List<int> arg2]);

void main() {
  assert(identical(F1, F2));
}
```

As far as I remember, the bug that started this series of tests was about the fact that `List<int>` and `List<ExtInt>` weren't canonicalized to the same object. This is the same thing for function types.

We just need to compare a function type that has extension types in every subterm against a function type where all those extension types have been erased manually, because that should be enough to verify that the erasure does take place during the canonicalization step (for instance, there's no need to test `void Function(ExtInt, ExtInt)` against `void Function(int, int)` and also `void Function(int, ExtInt)` against `void Function(ExtInt, int)` and so on, we just need "extension types everywhere" against "erased types everywhere").

Reply:
Thanks!